aten/src/ATen/native/Copy.cpp

```diff
@@ -124,21 +124,26 @@ static Tensor & copy_impl(Tensor & self, const Tensor & src, bool non_blocking)
   // 1. Memory Format for source and destination tensors is contiguous.
   // 2. Device for both the source and destination tensor is CPU.
   // 3. dtype conversion between FP32->FP16 and FP16->FP32.
+  // To avoid out of bounds memory access, this also checks whether there's
+  // enough self.numel() storage before copying src.numel() items, see
+  // fbgemm::Float16ToFloat_ref.
+  // https://github.com/pytorch/pytorch/issues/88543
   #ifdef USE_FBGEMM
     if (((self.dtype() == at::kFloat && src.dtype() == at::kHalf) ||
          (self.dtype() == at::kHalf && src.dtype() == at::kFloat)) &&
         (self.device().is_cpu() && src.device().is_cpu()) &&
         ((self.is_contiguous() && src.is_contiguous()) ||
```

Comment:
Note: this part of the conditional doesn't care about strides at all, which is why the behavior is different compared to TI. Moreover, there's no broadcasting support. I think the intention is to copy adjacent memory, but I feel like this might be confusing because the usual broadcasting rules are not honored for float16 (and only if you have fbgemm enabled).

There's an easy way to enable compatible behavior between fbgemm and TI if we add another case requiring shapes to match exactly, but this will reduce the amount of code that'll be executed by fbgemm. E.g., broadcasting cases will be handled by TI below, so probably not ideal.

Reply:
As discussed below, this should be a transparent optimisation, and we should make sure that we just take this path whenever it is equivalent to running the TI path.